akka-actor/src/main/scala/akka/util/ImmutableIntMap.scala

```diff
@@ -3,139 +3,118 @@
+    new ImmutableIntMap(newKvs, size)
+  }
 
-        new ImmutableIntMap(newKeys, newValues)
-      }
-    }
+  private[this] final def insert(key: Int, value: Int, index: Int): ImmutableIntMap = {
+    // insert the entry at the right position—keep the array sorted
+    val at = -(index + 1)
+    val newKvs = new Array[Int](kvs.length + 2)
+    System.arraycopy(kvs, 0, newKvs, 0, at)
+    newKvs(at) = key
+    newKvs(at + 1) = value
+    System.arraycopy(kvs, at, newKvs, at + 2, kvs.length - at)
+    new ImmutableIntMap(newKvs, size + 1)
   }
 
-  def remove(key: Int): ImmutableIntMap = {
-    val i = Arrays.binarySearch(keys, key)
+  final def remove(key: Int): ImmutableIntMap = {
+    val i = indexForKey(key)
     if (i >= 0) {
-      if (size == 1)
-        ImmutableIntMap.empty
-      else {
-        val newKeys = new Array[Int](size - 1)
-        System.arraycopy(keys, 0, newKeys, 0, i)
-        System.arraycopy(keys, i + 1, newKeys, i, keys.length - i - 1)
-
-        val newValues = new Array[Int](size - 1)
-        System.arraycopy(values, 0, newValues, 0, i)
-        System.arraycopy(values, i + 1, newValues, i, values.length - i - 1)
-
-        new ImmutableIntMap(newKeys, newValues)
-      }
-    } else
-      this
+      if (size > 1) {
+        val newKvs = new Array[Int](kvs.length - 2)
+        System.arraycopy(kvs, 0, newKvs, 0, i)
+        System.arraycopy(kvs, i + 2, newKvs, i, kvs.length - i - 2)
+        new ImmutableIntMap(newKvs, size - 1)
+      } else ImmutableIntMap.empty
+    } else this
   }
 
   /**
    * All keys
    */
-  def keysIterator: Iterator[Int] =
-    keys.iterator
+  final def keysIterator: Iterator[Int] =
+    if (size < 1) Iterator.empty
+    else Iterator.range(0, kvs.length - 1, 2).map(kvs.apply)
 
-  override def toString: String =
-    keysIterator.map(key ⇒ s"$key -> ${get(key)}").mkString("ImmutableIntMap(", ", ", ")")
+  override final def toString: String =
+    if (size < 1) "ImmutableIntMap()"
+    else Iterator.range(0, kvs.length - 1, 2).map(i ⇒ s"${kvs(i)} -> ${kvs(i + 1)}").mkString("ImmutableIntMap(", ", ", ")")
 
-  override def hashCode: Int = {
-    var result = HashCode.SEED
-    result = HashCode.hash(result, keys)
-    result = HashCode.hash(result, values)
-    result
-  }
+  override final def hashCode: Int = Arrays.hashCode(kvs)
 
-  override def equals(obj: Any): Boolean = obj match {
-    case other: ImmutableIntMap ⇒
-      if (other eq this) true
-      else if (size != other.size) false
-      else if (size == 0 && other.size == 0) true
-      else {
-        @tailrec def check(i: Int): Boolean = {
-          if (i < 0) true
-          else if (keys(i) == other.keys(i) && values(i) == other.values(i))
-            check(i - 1) // recur, next elem
-          else false
-        }
-        check(size - 1)
-      }
-    case _ ⇒ false
+  override final def equals(obj: Any): Boolean = obj match {
+    case other: ImmutableIntMap ⇒ Arrays.equals(kvs, other.kvs)
```

Comment:
Array.equals already compares the arrays for instance equality (which will be true if comparing something to itself) so no need to test that outside.
Also tests for equal lengths, so no need to test for size vs other.size.